File: CSW/JavaScript/220913_최소원_트럭.js
var fs = require('fs');
var input = fs.readFileSync(`C://Users/SONG/Desktop/pjt/Algorithm_Study_Jeans/CSW/JavaScript/example.txt`).toString().trim().split('\n');

let [N, W, L] = input[0].split(' ').map(e=>parseInt(e));
const truck = input[1].split(' ').map(e=>parseInt(e));
const bridge = Array.from({length:W}, ()=>0);

let weight = 0;
let minTime = 0;

while(N){
    minTime++;
    let outBridge = bridge.shift();
    if(outBridge) N--;
    if(!N)break;
    let onBridge = truck[0];
    weight =  weight + onBridge - outBridge;
    if(weight > L){
        weight -= onBridge;
        bridge.push(0);
    }else{
        let temp = truck.shift();
        truck.push(0);
        bridge.push(temp);
    }
}


console.log(minTime);

File: CSW/C++/220913_최소원_트럭.cpp
#include <iostream>
#include <queue>
using namespace std;

int N, W, L;
queue<int> bridge;
queue<int> truck;
int minTime;
int weight;

int main() {
	cin >> N >> W >> L;
	for (int i = 0; i < N; i++) {
		int num; cin >> num;
		truck.push(num);
	}

	for (int i = 0; i < W; i++) {
		bridge.push(0);
	}

	while (N) {
		minTime++;
		int outBridge = bridge.front();
		bridge.pop();
		if (outBridge) N--;
		if (!N) break;
		int onBridge = truck.front();
		weight = weight + onBridge - outBridge;
		if (weight > L) {
			weight -= onBridge;
			bridge.push(0);
		}
		else {
			bridge.push(onBridge);
			truck.pop();
			truck.push(0);
		}
	}

	cout << minTime;

	return 0;
}